private/buf/bufcli/bufcli.go

```diff
@@ -550,8 +550,36 @@ func NewConfig(container appflag.Container) (*bufapp.Config, error) {
 	return bufapp.NewConfig(container, externalConfig)
 }
 
-// NewRegistryProvider creates a new registryv1alpha1apiclient.Provider.
+func getTokenReader(container appflag.Container) func(string) (string, error) {
+	return func(address string) (string, error) {
+		token := container.Env(tokenEnvKey)
+		if token == "" {
+			machine, err := netrc.GetMachineForName(container, address)
+			if err != nil {
+				return "", fmt.Errorf("failed to read server password from netrc: %w", err)
+			}
+			if machine != nil {
+				token = machine.Password()
+			}
+		}
+		return token, nil
+	}
+}
+
+// NewRegistryProvider creates a new registryv1alpha1apiclient.Provider which uses a token reader to look
+// up the token in the container or in netrc based on the address of each individual client from the provider.
+// It is then set in the header of all outgoing requests from this provider
 func NewRegistryProvider(ctx context.Context, container appflag.Container) (registryv1alpha1apiclient.Provider, error) {
+	return newRegistryProviderWithOptions(container, bufapiclient.RegistryProviderWithTokenReader(getTokenReader(container)))
```

Comment:
Putting this up to gather some opinions on the best way to construct this.  The issue is that to look up the token, we need the address.  But at the time of creating the provider here, we don't have the address because that is done at each individual client level.  So, we have to pass a function to look it up when each client is created.

However, even doing it this way fails `lint` because now we are relying on `private/bufpkg` in the generated files to create the `WithTokenInterceptor` under the hood.  We could move the `bufconnect` package out of `private/bufpkg` but that's just a band-aid.  Hence this whole dilemma.

What are your thoughts here?  @bufdev @mfridman @amckinney

Reply:
Discussed offline - @smaye81 and I think we can get around this `bufpkg` dependency by explicitly passing in the interceptor that depends on the `bufpkg` details to the provider constructor. That way the generated code never interacts with the `bufpkg` types at all.

We'll see how that feels/works and go from there.

Reply:
I was able to remove the dependency by passing in interceptor details to the provider constructor but also using those details to construct the interceptor within the generated code.

It made sense to have the interceptor in the generated code since it is a Connect-specific thing.  I realized I just needed to pass in some parameters so that the interceptor could be instantiated.  I went with a `TokenConfig` to do so.  Let me know what everyone thinks.

Reply:
OK after speaking with @amckinney, did another pass at this using an auth interceptor provider pattern, which removes the need for the additional token config in generated code AND removes the dep on `bufconnect`.

Reply:
All things considered, this seems like the most sane solution. Good job @smaye81 @amckinney 

In hindsight, we probably should have got rid of the provider abstraction, but that's not here nor there.